blivet/devicelibs/lvm.py

```diff
@@ -65,78 +65,90 @@
+    device_string = ""
+
+    if not HAVE_LVMDEVICES:
+        # now explicitly "accept" all LVM devices
+        for device in _lvm_devices:
+            device_string += "\"a|%s$|\"," % device
+
+        # now add all devices to the "reject" filter
+        device_string += "\"r|.*|\""
 
-    if filter_string:
-        filter_string = "filter=[%s]" % filter_string.strip(",")
+        filter_string = "filter=[%s]" % device_string
 
-    # XXX consider making /tmp/blivet.lvm.XXXXX, writing an lvm.conf there, and
-    #     setting LVM_SYSTEM_DIR
-    devices_string = 'preferred_names=["^/dev/mapper/", "^/dev/md/", "^/dev/sd"]'
-    if filter_string:
-        devices_string += " %s" % filter_string
+        config_string = " devices { %s } " % filter_string
+    else:
+        config_string = " "
 
-    # devices_string can have (inside the brackets) "dir", "scan",
-    # "preferred_names", "filter", "cache_dir", "write_cache_state",
-    # "types", "sysfs_scan", "md_component_detection".  see man lvm.conf.
-    config_string = " devices { %s } " % (devices_string)  # strings can be added
     if not flags.lvm_metadata_backup:
         config_string += "backup {backup=0 archive=0} "
-    if flags.debug:
-        config_string += "log {level=7 file=/tmp/lvm.log syslog=0}"
+    config_string += "log {level=7 file=/tmp/lvm.log syslog=0}"
 
     blockdev.lvm.set_global_config(config_string)
 
 
+def _set_lvm_devices():
+    if not HAVE_LVMDEVICES:
+        return
+
+    blockdev.lvm_set_devices_filter(list(_lvm_devices))
+
+
 def needs_config_refresh(fn):
     if not availability.BLOCKDEV_LVM_PLUGIN.available:
         return lambda *args, **kwargs: None
 
     def fn_with_refresh(*args, **kwargs):
         ret = fn(*args, **kwargs)
         _set_global_config()
+        _set_lvm_devices()
         return ret
 
     return fn_with_refresh
 
 
 @needs_config_refresh
-def lvm_cc_addFilterRejectRegexp(regexp):
-    """ Add a regular expression to the --config string."""
-    log.debug("lvm filter: adding %s to the reject list", regexp)
-    config_args_data["filterRejects"].add(regexp)
+def lvm_devices_add(path):
+    """ Add a device (PV) to the list of devices LVM is allowed to use """
+    log.debug("lvm filter: device %s added to the list of allowed devices")
+    _lvm_devices.add(path)
 
 
 @needs_config_refresh
-def lvm_cc_removeFilterRejectRegexp(regexp):
-    """ Remove a regular expression from the --config string."""
-    log.debug("lvm filter: removing %s from the reject list", regexp)
+def lvm_devices_remove(path):
+    """ Remove a device (PV) to the list of devices LVM is allowed to use """
+    log.debug("lvm filter: device %s removed from the list of allowed devices")
     try:
-        config_args_data["filterRejects"].remove(regexp)
+        _lvm_devices.remove(path)
     except KeyError:
-        log.debug("%s wasn't in the reject list", regexp)
+        log.debug("%s wasn't in the devices list", path)
```

Comment:
Another (quieter) approach would be to check for `path` and only emit the log message if it is indeed being added/removed as requested -- IOW, if the call is a no-op nothing would be logged. Maybe that's less clear for debugging the devices list, but less noise (at least the way we use the old/current filter) in general.

Reply:
The excessive was helpful when debugging some of the filtering issue. Ideally there should be zero no-op calls here. We could change this in the future when (if) we decide to drop support for the "old style" lvm.conf filtering and the device file based filtering is mature enough.